crate/operator/handlers/handle_restore_backup.py

```diff
@@ -0,0 +1,273 @@
+# software solely pursuant to the terms of the relevant commercial agreement.
+
+import datetime
+import hashlib
+import logging
+from typing import List
+
+import kopf
+
+from crate.operator.config import config
+from crate.operator.handlers.handle_update_cratedb import get_backoff
+from crate.operator.operations import (
+    AfterClusterUpdateSubHandler,
+    BeforeClusterUpdateSubHandler,
+    get_cratedb_resource,
+)
+from crate.operator.restore_backup import (
+    AfterRestoreBackupSubHandler,
+    BeforeRestoreBackupSubHandler,
+    ResetSnapshotSubHandler,
+    RestoreBackupSubHandler,
+    RestoreSystemUserPasswordSubHandler,
+    SendSuccessNotificationSubHandler,
+    ensure_no_restore_in_progress,
+    get_crash_pod_name,
+    get_crash_scheme,
+)
+from crate.operator.utils.notifications import FlushNotificationsSubHandler
+
+CLUSTER_RESTORE_FIELD_ID = "cluster_restore/spec.cluster.restoreSnapshot"
+
+
+async def restore_backup(
+    namespace: str,
+    name: str,
+    diff: kopf.Diff,
+    new: kopf.Body,
+    patch: kopf.Patch,
+    status: kopf.Status,
+    started: datetime.datetime,
+    logger: logging.Logger,
+):
+    context = status.get(CLUSTER_RESTORE_FIELD_ID)
+    hash_string = str(diff) + str(started)
+    change_hash = hashlib.md5(hash_string.encode("utf-8")).hexdigest()
+    if not context:
+        context = {"ref": change_hash}
+    elif context.get("ref", "") != change_hash:
+        context["ref"] = change_hash
+
+    cratedb = await get_cratedb_resource(namespace, name)
+    snapshot = new["snapshot"]
+    tables = new.get("tables", ["all"])
+
+    scheme = get_crash_scheme(cratedb)
+    pod_name = get_crash_pod_name(cratedb, name)
+    repository = f"restore_backup_{int(started.timestamp())}"
+
+    # If there is a restore operation in progress, we do not do
+    # anything else until it is finished.
+    await ensure_no_restore_in_progress(
+        namespace, name, snapshot, pod_name, scheme, logger
+    )
+
+    depends_on: List[str] = []
+
+    register_before_restore_handlers(
+        namespace, name, patch, change_hash, context, depends_on
+    )
+
+    register_restore_handlers(
+        namespace,
+        name,
+        change_hash,
+        context,
+        depends_on,
+        repository,
+        snapshot,
+        tables,
+    )
+
+    register_after_restore_handlers(
+        namespace,
+        name,
+        status,
+        change_hash,
+        context,
+        depends_on,
+        repository,
+        snapshot,
+    )
+
+    kopf.register(
+        fn=FlushNotificationsSubHandler(
+            namespace,
+            name,
+            change_hash,
+            context,
+            depends_on=depends_on.copy(),
+            run_on_dep_failures=True,
+        )(),
+        id="flush_notifications",
+        backoff=get_backoff(),
+    )
+    depends_on.append(f"{CLUSTER_RESTORE_FIELD_ID}/flush_notifications")
+
+    # This needs to be the last subhandler, otherwise the operation does
```

Comment:
Like this, it seems to finish properly. All the annotations with the subhandler status are deleted.

```
Handler 'cluster_restore/spec.cluster.restoreSnapshot' succeeded.
Updating is processed: 2 succeeded; 0 failed.
```

I think 1 for the initial update and the 2nd for the update to `snapshot: ""` which does not really run because `when=is_valid_snapshot` returns `False`.